File: sdks/ts/memory/src/memory/hygiene.ts
// SPDX-License-Identifier: Apache-2.0

import { ErrNotFound } from '../store/errors.js'
import type { Batch, Store } from '../store/index.js'
import { type Path, lastSegment } from '../store/path.js'
import { type Frontmatter, buildFrontmatter, parseFrontmatter } from './frontmatter.js'
import { scopePrefix } from './paths.js'
import type { Scope } from './types.js'

export const DEFAULT_RETIRED_AGE_DAYS = 30

export type MemoryHygieneArgs = {
  readonly scope?: Scope
  readonly actorId?: string
  readonly retiredAgeDays?: number
  readonly apply?: boolean
  readonly now?: Date | string
}

export type RunMemoryHygieneOptions = MemoryHygieneArgs & {
  readonly store: Store
  readonly scope: Scope
  readonly actorId: string
}

export type MemoryHygieneTopic = {
  readonly name: string
  readonly description: string
  readonly type: string
  readonly path: Path
  readonly created?: string
  readonly modified?: string
  readonly tags: readonly string[]
  readonly confidence?: string
  readonly source?: string
  readonly scope: Scope
}

export type MemoryContradictionGroup = {
  readonly key: string
  readonly keyReason: 'claim_key' | 'name'
  readonly scope: Scope
  readonly actorId: string
  readonly members: readonly MemoryHygieneTopic[]
  readonly canonical?: Path
}

export type MemoryAgingRetirement = {
  readonly path: Path
  readonly ageMs: number
}

export type MemoryHygieneReport = {
  readonly contradictions: readonly MemoryContradictionGroup[]
  readonly agingRetired: readonly MemoryAgingRetirement[]
  readonly errors: readonly string[]
}

type TopicRecord = MemoryHygieneTopic & {
  readonly frontmatter: Frontmatter
  readonly content: string
  readonly body: string
}

export const runMemoryHygiene = async (
  opts: RunMemoryHygieneOptions,
): Promise<MemoryHygieneReport> => {
  const now = coerceDate(opts.now) ?? new Date()
  const retiredAgeDays =
    opts.retiredAgeDays === undefined || opts.retiredAgeDays <= 0
      ? DEFAULT_RETIRED_AGE_DAYS
      : opts.retiredAgeDays
  const maxAgeMs = retiredAgeDays * 24 * 60 * 60 * 1000
  const prefix = scopePrefix(opts.scope, opts.actorId)
  const errors: string[] = []
  const topics = await listTopicRecords(opts.store, prefix, opts.scope, errors)

  const contradictions = detectContradictions(topics, opts.scope, opts.actorId)
  const agingRetired = detectAgingRetirements(topics, now, maxAgeMs)

  if (opts.apply === true && (contradictions.length > 0 || agingRetired.length > 0)) {
    await applyHygiene(opts.store, {
      scope: opts.scope,
      actorId: opts.actorId,
      now,
      contradictions,
      agingRetired,
      errors,
    })
  }

  return {
    contradictions:
      opts.apply === true
        ? contradictions
        : contradictions.map(({ canonical: _canonical, ...group }) => group),
    agingRetired,
    errors,
  }
}

const detectContradictions = (
  topics: readonly TopicRecord[],
  scope: Scope,
  actorId: string,
): MemoryContradictionGroup[] => {
  const groupsByClaim = new Map<string, TopicRecord[]>()
  const groupsByName = new Map<string, TopicRecord[]>()

  for (const topic of topics) {
    if (topic.frontmatter.superseded_by || topic.frontmatter.retired === true) continue
    if (topic.frontmatter.claim_key) {
      appendGroup(groupsByClaim, topic.frontmatter.claim_key, topic)
    }
    if (topic.frontmatter.name) {
      appendGroup(groupsByName, topic.frontmatter.name.trim().toLowerCase(), topic)
    }
  }

  return [
    ...emitContradictionGroups(groupsByClaim, 'claim_key', scope, actorId),
    ...emitContradictionGroups(groupsByName, 'name', scope, actorId),
  ]
}

const appendGroup = (groups: Map<string, TopicRecord[]>, key: string, topic: TopicRecord): void => {
  const existing = groups.get(key)
  if (existing === undefined) {
    groups.set(key, [topic])
    return
  }
  existing.push(topic)
}

const emitContradictionGroups = (
  groups: ReadonlyMap<string, readonly TopicRecord[]>,
  keyReason: MemoryContradictionGroup['keyReason'],
  scope: Scope,
  actorId: string,
): MemoryContradictionGroup[] => {
  const out: MemoryContradictionGroup[] = []
  const keys = [...groups.keys()].sort()
  for (const key of keys) {
    const members = uniqueTopics(groups.get(key) ?? [])
    if (members.length < 2) continue
    out.push({
      key,
      keyReason,
      scope,
      actorId,
      members,
    })
  }
  return out
}

const detectAgingRetirements = (
  topics: readonly TopicRecord[],
  now: Date,
  maxAgeMs: number,
): MemoryAgingRetirement[] => {
  const out: MemoryAgingRetirement[] = []
  const threshold = now.getTime() - maxAgeMs
  for (const topic of topics) {
    if (!topic.frontmatter.superseded_by || topic.frontmatter.retired === true) continue
    const modified = parseDate(topic.frontmatter.modified)
    if (modified === undefined || modified.getTime() >= threshold) continue
    out.push({
      path: topic.path,
      ageMs: now.getTime() - modified.getTime(),
    })
  }
  return out
}

const applyHygiene = async (
  store: Store,
  input: {
    readonly scope: Scope
    readonly actorId: string
    readonly now: Date
    readonly contradictions: MemoryContradictionGroup[]
    readonly agingRetired: readonly MemoryAgingRetirement[]
    readonly errors: string[]
  },
): Promise<void> => {
  try {
    await store.batch(
      { reason: `memory:hygiene ${input.scope}${projectSuffix(input)}` },
      async (b) => {
        for (let index = 0; index < input.contradictions.length; index++) {
          const group = input.contradictions[index]
          if (group === undefined) continue
          const canonical = pickCanonical(group.members)
          if (canonical === undefined) continue
          input.contradictions[index] = { ...group, canonical: canonical.path }
          const canonicalFile = lastSegment(canonical.path)
          for (const member of group.members) {
            if (member.path === canonical.path) continue
            try {
              await stampSupersededBy(b, member.path, canonicalFile)
            } catch (err) {
              input.errors.push(
                `stamp ${member.path}: ${err instanceof Error ? err.message : String(err)}`,
              )
            }
          }
        }

        for (const retirement of input.agingRetired) {
          try {
            const existing = (await b.read(retirement.path)).toString('utf8')
            const reason = `auto-retired by hygiene: superseded for ${formatAge(retirement.ageMs)}`
            const stamped = retireFrontmatterInPlace(existing, input.now, reason)
            await b.write(retirement.path, Buffer.from(stamped, 'utf8'))
          } catch (err) {
            input.errors.push(
              `retire ${retirement.path}: ${err instanceof Error ? err.message : String(err)}`,
            )
          }
        }
      },
    )
  } catch (err) {
    input.errors.push(`hygiene batch failed: ${err instanceof Error ? err.message : String(err)}`)
  }
}

const listTopicRecords = async (
  store: Store,
  prefix: Path,
  scope: Scope,
  errors: string[],
): Promise<TopicRecord[]> => {
  let entries: Awaited<ReturnType<Store['list']>>
  try {
    entries = await store.list(prefix, { recursive: true, includeGenerated: true })
  } catch (err) {
    if (err instanceof ErrNotFound) return []
    errors.push(`listing ${prefix}: ${err instanceof Error ? err.message : String(err)}`)
    return []
  }

  const topics: TopicRecord[] = []
  for (const entry of entries) {
    if (entry.isDir) continue
    const filename = lastSegment(entry.path)
    if (!filename.endsWith('.md') || filename === 'MEMORY.md') continue
    try {
      const content = (await store.read(entry.path)).toString('utf8')
      const parsed = parseFrontmatter(content)
      const topicName = parsed.frontmatter.name ?? filename.replace(/\.md$/, '')
      topics.push({
        name: topicName,
        description: parsed.frontmatter.description ?? '',
        type: parsed.frontmatter.type ?? '',
        path: entry.path,
        tags: parsed.frontmatter.tags ?? [],
        scope,
        frontmatter: parsed.frontmatter,
        content,
        body: parsed.body,
        ...(parsed.frontmatter.created !== undefined
          ? { created: parsed.frontmatter.created }
          : {}),
        ...(parsed.frontmatter.modified !== undefined
          ? { modified: parsed.frontmatter.modified }
          : {}),
        ...(parsed.frontmatter.confidence !== undefined
          ? { confidence: parsed.frontmatter.confidence }
          : {}),
        ...(parsed.frontmatter.source !== undefined ? { source: parsed.frontmatter.source } : {}),
      })
    } catch (err) {
      errors.push(`reading ${entry.path}: ${err instanceof Error ? err.message : String(err)}`)
    }
  }

  return topics
}

const pickCanonical = (members: readonly MemoryHygieneTopic[]): MemoryHygieneTopic | undefined => {
  return [...members].sort((left, right) => {
    const confidenceDelta = confidenceRank(right.confidence) - confidenceRank(left.confidence)
    if (confidenceDelta !== 0) return confidenceDelta
    const leftModified = parseDate(left.modified)?.getTime() ?? 0
    const rightModified = parseDate(right.modified)?.getTime() ?? 0
    if (leftModified !== rightModified) return rightModified - leftModified
    return left.path.localeCompare(right.path)
  })[0]
}

const stampSupersededBy = async (batch: Batch, path: Path, newFile: string): Promise<void> => {
  const content = (await batch.read(path)).toString('utf8')
  if (!hasFrontmatter(content)) return
  const parsed = parseFrontmatter(content)
  const next = buildNoteContent(
    {
      ...parsed.frontmatter,
      superseded_by: newFile,
    },
    parsed.body,
  )
  await batch.write(path, Buffer.from(next, 'utf8'))
}

const retireFrontmatterInPlace = (content: string, now: Date, reason: string): string => {
  if (!hasFrontmatter(content)) return content
  const parsed = parseFrontmatter(content)
  if (parsed.frontmatter.retired === true) return content
  return buildNoteContent(
    {
      ...parsed.frontmatter,
      retired: true,
      retired_on: formatDateOnly(now),
      ...(reason !== '' ? { retired_reason: reason } : {}),
    },
    parsed.body,
  )
}

const buildNoteContent = (frontmatter: Frontmatter, body: string): string => {
  const built = buildFrontmatter(frontmatter)
  const trimmedBody = body.trim()
  return trimmedBody === '' ? `${built}\n` : `${built}\n${trimmedBody}\n`
}

const uniqueTopics = (topics: readonly TopicRecord[]): MemoryHygieneTopic[] => {
  const out: MemoryHygieneTopic[] = []
  const seen = new Set<string>()
  for (const topic of topics) {
    if (seen.has(topic.path)) continue
    seen.add(topic.path)
    out.push(topic)
  }
  return out
}

const confidenceRank = (value: string | undefined): number => {
  switch (value?.trim().toLowerCase()) {
    case 'high':
      return 3
    case 'medium':
      return 2
    case 'low':
      return 1
    default:
      return 0
  }
}

const parseDate = (value: string | undefined): Date | undefined => {
  if (value === undefined || value.trim() === '') return undefined
  const parsed = new Date(value)
  return Number.isNaN(parsed.getTime()) ? undefined : parsed
}

const coerceDate = (value: Date | string | undefined): Date | undefined => {
  if (value === undefined) return undefined
  if (value instanceof Date) return Number.isNaN(value.getTime()) ? undefined : value
  return parseDate(value)
}

const hasFrontmatter = (content: string): boolean => content.split('\n')[0]?.trim() === '---'

const formatDateOnly = (date: Date): string => date.toISOString().slice(0, 10)

const formatAge = (ageMs: number): string => {
  const days = Math.round(ageMs / (24 * 60 * 60 * 1000))
  return `${days * 24}h0m0s`
}

const projectSuffix = (input: { readonly scope: Scope; readonly actorId: string }): string =>
  input.scope === 'project' ? `/${input.actorId}` : ''

File: go/memory/hygiene.go
// SPDX-License-Identifier: Apache-2.0

package memory

import (
	"context"
	"errors"
	"fmt"
	"path"
	"sort"
	"strings"
	"time"

	"github.com/jeffs-brain/memory/go/brain"
)

// DefaultRetiredAgeDays is the default age before a superseded memory
// file is stamped with retired: true by the hygiene pass.
const DefaultRetiredAgeDays = 30

// HygieneOptions tunes a single hygiene pass.
type HygieneOptions struct {
	// RetiredAgeDays controls when superseded entries are soft-retired.
	// Zero or negative means use DefaultRetiredAgeDays.
	RetiredAgeDays int

	// Apply controls whether the pass writes changes. When false the pass
	// populates the report but never mutates the store.
	Apply bool

	// Now pins time for tests and reproducible reports. Zero means time.Now.
	Now time.Time
}

// HygieneReport summarises what the contradiction and aging pass found.
type HygieneReport struct {
	Contradictions []ContradictionGroup
	AgingRetired   []AgingRetirement
	Errors         []string
}

// ContradictionGroup describes live memory files that cover the same
// topic. When Apply is set, every entry except Canonical is stamped with
// superseded_by pointing at Canonical.
type ContradictionGroup struct {
	Key       string
	KeyReason string
	Scope     string
	Project   string
	Members   []TopicFile
	Canonical brain.Path
}

// AgingRetirement describes a superseded file that should be soft-retired.
type AgingRetirement struct {
	Path brain.Path
	Age  time.Duration
}

// RunHygiene performs a contradiction and aging pass across every memory
// scope. Dry-run mode reports intended changes. Apply mode writes one
// batch per scope.
func (c *Consolidator) RunHygiene(ctx context.Context, opts HygieneOptions) (*HygieneReport, error) {
	c.mu.Lock()
	if c.inProgress {
		c.mu.Unlock()
		return nil, fmt.Errorf("consolidation already in progress")
	}
	c.inProgress = true
	c.mu.Unlock()
	defer func() {
		c.mu.Lock()
		c.inProgress = false
		c.mu.Unlock()
	}()

	report := &HygieneReport{}

	maxAgeDays := opts.RetiredAgeDays
	if maxAgeDays <= 0 {
		maxAgeDays = DefaultRetiredAgeDays
	}
	maxAge := time.Duration(maxAgeDays) * 24 * time.Hour
	now := opts.Now
	if now.IsZero() {
		now = time.Now()
	}

	prefixes := c.scopePrefixes(ctx)
	for _, prefix := range prefixes {
		c.detectContradictionsForPrefix(ctx, prefix, report)
		c.ageRetireSupersededForPrefix(ctx, prefix, now, maxAge, report)
	}

	if !opts.Apply {
		for i := range report.Contradictions {
			report.Contradictions[i].Canonical = ""
		}
		return report, nil
	}

	for _, prefix := range prefixes {
		c.applyHygieneForPrefix(ctx, prefix, report, now)
	}

	return report, nil
}

func (c *Consolidator) detectContradictionsForPrefix(ctx context.Context, prefix brain.Path, report *HygieneReport) {
	scope, project := hygieneScopeForPrefix(prefix)

	entries, err := c.mem.store.List(ctx, prefix, brain.ListOpts{IncludeGenerated: true})
	if err != nil {
		if !errors.Is(err, brain.ErrNotFound) {
			report.Errors = append(report.Errors, fmt.Sprintf("listing %s: %s", prefix, err))
		}
		return
	}

	groupsByClaim := map[string][]TopicFile{}
	groupsByName := map[string][]TopicFile{}

	for _, e := range entries {
		if e.IsDir {
			continue
		}
		base := path.Base(string(e.Path))
		if !strings.HasSuffix(base, ".md") || strings.EqualFold(base, "MEMORY.md") {
			continue
		}
		data, err := c.mem.store.Read(ctx, e.Path)
		if err != nil {
			continue
		}
		fm, _ := ParseFrontmatter(string(data))
		if fm.SupersededBy != "" || fm.Retired {
			continue
		}
		topic := TopicFile{
			Name:        fm.Name,
			Description: fm.Description,
			Type:        fm.Type,
			Path:        e.Path,
			Created:     fm.Created,
			Modified:    fm.Modified,
			Tags:        fm.Tags,
			Confidence:  fm.Confidence,
			Source:      fm.Source,
			Scope:       scope,
		}

		if fm.ClaimKey != "" {
			groupsByClaim[fm.ClaimKey] = append(groupsByClaim[fm.ClaimKey], topic)
		}
		if fm.Name != "" {
			key := strings.ToLower(strings.TrimSpace(fm.Name))
			groupsByName[key] = append(groupsByName[key], topic)
		}
	}

	emitContradictionGroups(report, groupsByClaim, "claim_key", scope, project)
	emitContradictionGroups(report, groupsByName, "name", scope, project)
}

func emitContradictionGroups(report *HygieneReport, groupsByKey map[string][]TopicFile, reason, scope, project string) {
	keys := make([]string, 0, len(groupsByKey))
	for k := range groupsByKey {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		members := groupsByKey[k]
		if len(members) < 2 {
			continue
		}
		seen := map[brain.Path]bool{}
		unique := make([]TopicFile, 0, len(members))
		for _, m := range members {
			if seen[m.Path] {
				continue
			}
			seen[m.Path] = true
			unique = append(unique, m)
		}
		if len(unique) < 2 {
			continue
		}
		report.Contradictions = append(report.Contradictions, ContradictionGroup{
			Key:       k,
			KeyReason: reason,
			Scope:     scope,
			Project:   project,
			Members:   unique,
		})
	}
}

func pickCanonical(members []TopicFile) TopicFile {
	if len(members) == 0 {
		return TopicFile{}
	}
	confidenceRank := func(s string) int {
		switch strings.ToLower(strings.TrimSpace(s)) {
		case "high":
			return 3
		case "medium":
			return 2
		case "low":
			return 1
		}
		return 0
	}
	sorted := make([]TopicFile, len(members))
	copy(sorted, members)
	sort.SliceStable(sorted, func(i, j int) bool {
		ri, rj := confidenceRank(sorted[i].Confidence), confidenceRank(sorted[j].Confidence)
		if ri != rj {
			return ri > rj
		}
		ti := parseHygieneModified(sorted[i].Modified)
		tj := parseHygieneModified(sorted[j].Modified)
		if !ti.Equal(tj) {
			return ti.After(tj)
		}
		return string(sorted[i].Path) < string(sorted[j].Path)
	})
	return sorted[0]
}

func parseHygieneModified(s string) time.Time {
	if s == "" {
		return time.Time{}
	}
	if t, err := time.Parse(time.RFC3339, s); err == nil {
		return t
	}
	return time.Time{}
}

func (c *Consolidator) ageRetireSupersededForPrefix(ctx context.Context, prefix brain.Path, now time.Time, maxAge time.Duration, report *HygieneReport) {
	threshold := now.Add(-maxAge)

	entries, err := c.mem.store.List(ctx, prefix, brain.ListOpts{IncludeGenerated: true})
	if err != nil {
		if !errors.Is(err, brain.ErrNotFound) {
			report.Errors = append(report.Errors, fmt.Sprintf("listing %s: %s", prefix, err))
		}
		return
	}
	for _, e := range entries {
		if e.IsDir {
			continue
		}
		base := path.Base(string(e.Path))
		if !strings.HasSuffix(base, ".md") || strings.EqualFold(base, "MEMORY.md") {
			continue
		}
		data, err := c.mem.store.Read(ctx, e.Path)
		if err != nil {
			continue
		}
		fm, _ := ParseFrontmatter(string(data))
		if fm.SupersededBy == "" || fm.Retired {
			continue
		}
		modified := parseHygieneModified(fm.Modified)
		if modified.IsZero() || !modified.Before(threshold) {
			continue
		}
		report.AgingRetired = append(report.AgingRetired, AgingRetirement{
			Path: e.Path,
			Age:  now.Sub(modified),
		})
	}
}

func (c *Consolidator) applyHygieneForPrefix(ctx context.Context, prefix brain.Path, report *HygieneReport, now time.Time) {
	scope, project := hygieneScopeForPrefix(prefix)

	var groupsHere []*ContradictionGroup
	for i := range report.Contradictions {
		g := &report.Contradictions[i]
		if g.Scope == scope && g.Project == project {
			groupsHere = append(groupsHere, g)
		}
	}
	var retireHere []*AgingRetirement
	for i := range report.AgingRetired {
		a := &report.AgingRetired[i]
		if strings.HasPrefix(string(a.Path), string(prefix)+"/") {
			retireHere = append(retireHere, a)
		}
	}
	if len(groupsHere) == 0 && len(retireHere) == 0 {
		return
	}

	err := c.mem.store.Batch(ctx, brain.BatchOptions{Reason: "memory:hygiene " + scope + hygieneProjectSuffix(project)}, func(b brain.Batch) error {
		for _, g := range groupsHere {
			canonical := pickCanonical(g.Members)
			g.Canonical = canonical.Path
			canonicalFile := path.Base(string(canonical.Path))
			for _, m := range g.Members {
				if m.Path == canonical.Path {
					continue
				}
				if _, err := stampSupersededBy(ctx, b, m.Path, canonicalFile); err != nil {
					report.Errors = append(report.Errors, fmt.Sprintf("stamp %s: %s", m.Path, err))
				}
			}
		}
		for _, a := range retireHere {
			existing, err := b.Read(ctx, a.Path)
			if err != nil {
				report.Errors = append(report.Errors, fmt.Sprintf("read %s: %s", a.Path, err))
				continue
			}
			reason := fmt.Sprintf("auto-retired by hygiene: superseded for %s", a.Age.Round(24*time.Hour))
			stamped := retireFrontmatterInPlace(existing, now, reason)
			if err := b.Write(ctx, a.Path, stamped); err != nil {
				report.Errors = append(report.Errors, fmt.Sprintf("retire %s: %s", a.Path, err))
			}
		}
		return nil
	})
	if err != nil {
		report.Errors = append(report.Errors, fmt.Sprintf("hygiene batch %s: %s", prefix, err))
	}
}

func retireFrontmatterInPlace(existing []byte, now time.Time, reason string) []byte {
	content := string(existing)
	lines := strings.Split(content, "\n")
	if len(lines) < 2 || strings.TrimSpace(lines[0]) != "---" {
		return existing
	}
	closeIdx := -1
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == "---" {
			closeIdx = i
			break
		}
	}
	if closeIdx < 0 {
		return existing
	}
	for i := 1; i < closeIdx; i++ {
		if strings.HasPrefix(strings.TrimSpace(lines[i]), "retired:") {
			return existing
		}
	}

	insert := []string{
		"retired: true",
		"retired_on: " + now.UTC().Format("2006-01-02"),
	}
	if reason != "" {
		insert = append(insert, "retired_reason: "+reason)
	}
	merged := make([]string, 0, len(lines)+len(insert))
	merged = append(merged, lines[:closeIdx]...)
	merged = append(merged, insert...)
	merged = append(merged, lines[closeIdx:]...)
	return []byte(strings.Join(merged, "\n"))
}

func hygieneScopeForPrefix(prefix brain.Path) (scope, project string) {
	s := string(prefix)
	if s == "memory/global" {
		return "global", ""
	}
	if strings.HasPrefix(s, "memory/project/") {
		return "project", strings.TrimPrefix(s, "memory/project/")
	}
	return "", ""
}

func hygieneProjectSuffix(project string) string {
	if project == "" {
		return ""
	}
	return "/" + project
}
